src/index.js

```diff
@@ -8,12 +8,24 @@ export default function register(Component, tagName, propNames, options) {
 			options && options.shadow ? inst.attachShadow({ mode: 'open' }) : inst;
 		return inst;
 	}
+
 	PreactElement.prototype = Object.create(HTMLElement.prototype);
 	PreactElement.prototype.constructor = PreactElement;
 	PreactElement.prototype.connectedCallback = connectedCallback;
 	PreactElement.prototype.attributeChangedCallback = attributeChangedCallback;
 	PreactElement.prototype.disconnectedCallback = disconnectedCallback;
 
+	// b \ a
+	// take all Component properties and remove existing PreactElement properties
+	const a = new Set(Object.getOwnPropertyNames(PreactElement.prototype));
+	const b = new Set(Object.getOwnPropertyNames(Component.prototype));
+	const diff = Array.from(new Set([...b].filter((x) => !a.has(x))));
+
+	PreactElement.prototype = diff.reduce((acc, propertyName) => {
+		acc[propertyName] = Component.prototype[propertyName];
```

Comment:
Likely need to invoke this on the corresponding method - right now the value of `this` within component methods invoked from the Custom Element will be the Custom Element rather than the Component instance.  It took me a while to find the (hilariously obscure) path, but it's `this._root.__k.__k[0].__k[0].__c` (dom -> preact root --> first child (ContextProvider) --> first child (your component) --> component instance).

Reply:
I did try to go down the `this._root.__k.__k[0].__k[0].__c` path first. However, that path gets populated only after DOM insertion which makes custom element creation via `document.createElement(...)` impossible.

Reply:
Hm, it seems Preact's component `constructor()` doesn't get called on `document.createElement('x-hello')`.

```js
class CounterPreact extends Component {
  constructor() {
    super()
    console.log('here')
  }
}

register(CounterPreact, 'x-counter-preact')

// ...

document.createElement('x-counter-preact') // no output
```

And now the same in vanilla JS:

```js
class CounterPlain extends HTMLElement {
  constructor() {
    super()
    console.log('here')
  }
}

customElements.define('x-counter-plain', CounterPlain)

// ...

document.createElement('x-counter-plain') // outputs here
```

---

This can be problematic when methods rely on `constructor()`.

```js
class Counter extends Component {
  constructor() {
    super()
    this.c = 0
  }

  count() {
    this.c++
    return this.c
  }
}

register(Counter, 'x-counter')

// ...

const counter = document.createElement('x-counter')
counter.count() // NaN
```

@developit @JoviDeCroock any thoughts?

Reply:
fwiw Preact's component constructor not being called when `document.createElement` is used doesn't seem related to the changes introduced via this PR.

I wonder if that should be it's own Issue and we could re-open this PR as I think the approach taken here is sound and a valuable feature for the community.